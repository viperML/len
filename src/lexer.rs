use chumsky::extra::ParserExtra;
use chumsky::input::StrInput;

use chumsky::text::Char;

use crate::Int;
use chumsky::prelude::*;
use chumsky::Parser;

#[derive(Debug, Clone, PartialEq, PartialOrd)]
pub enum TokenKind<'src> {
    Number(Int),
    Bind,
    String(&'src str),
    Ident(&'src str),
    Symbol(&'src str),
    RightParenthesis,
    LeftParenthesis,
}

#[derive(Debug)]
pub struct Token<'src> {
    pub kind: TokenKind<'src>,
    pub span: SimpleSpan,
}

fn is_reserved_char(c: &char) -> bool {
    r#"(),;[]`{}_:"'"#.chars().any(|reserved| reserved == c.to_char())
}

#[must_use]
fn symbol<'a, I: StrInput<'a, C>, C: Char, E: ParserExtra<'a, I>>(
) -> impl Parser<'a, I, &'a C::Str, E> + Copy + Clone {
    let f = |c: &C| {
        let c = c.to_char();
        !c.is_whitespace() && !c.is_alphanumeric() && !is_reserved_char(&c)
    };

    any().filter(f).then(any().filter(f).repeated()).to_slice()
}

pub type LexerI<'a> = &'a str;
pub type LexerO<'a> = Vec<Token<'a>>;

#[must_use]
pub fn lexer<'s, E: ParserExtra<'s, LexerI<'s>>>() -> impl Parser<'s, LexerI<'s>, LexerO<'s>, E> {
    let number = text::int(10).map(|s: &str| TokenKind::Number(s.parse().unwrap()));

    let string = any()
        .filter(|c| *c != '"')
        .repeated()
        .to_slice()
        .map(TokenKind::String)
        .delimited_by(just('"'), just('"'));

    let right_parens = just(')').map(|_| TokenKind::RightParenthesis);
    let left_parens = just('(').map(|_| TokenKind::LeftParenthesis);

    let symbol = symbol().to_slice().map(TokenKind::Symbol);

    let ident = chumsky::text::unicode::ident().map(TokenKind::Ident);

    number
        .or(right_parens)
        .or(left_parens)
        .or(symbol)
        .or(string)
        .or(ident)
        .padded()
        .map_with(|t: TokenKind, e| Token {
            kind: t,
            span: e.span(),
        })
        .repeated()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use chumsky::error;
    use insta::assert_debug_snapshot;
    use rstest::rstest;
    use tracing_test::traced_test;

    type TestExtra = extra::Err<error::Cheap>;

    #[rstest]
    #[traced_test]
    fn test_symbol(
        #[values(
            //-
            " + ",
            " + ",
            " +2",
            " ++ ",
            " // ",
            " - - - -- - - ",
            " () , . ",
        )]
        input: &str,
    ) {
        let p = symbol::<_, _, TestExtra>()
            .padded()
            .repeated()
            .collect::<Vec<_>>();

        assert_debug_snapshot!(input, (input, p.parse(input)));
    }

    #[rstest]
    #[traced_test]
    fn test_lexer(
        #[values(
            ("empty", ""),
            ("int", "1 23 313 1"),
            ("symbols", "+ == != -"),
            ("symbols_split", r#"+3+1--3//3&(s++)"#),
            ("string", r#" "foo" "bar" "#),
            ("bad s", r#" "foo "#),
            ("parens", r#"(12 +23)()("foo")(1+1)"#),
            ("ident", "foo bar foo_bar foo-bar (foo+1)"),
        )]
        input: (&str, &str),
    ) {
        let p = lexer::<TestExtra>().padded();

        assert_debug_snapshot!(input.0, (input.0, input.1, p.parse(input.1)));
    }
}
