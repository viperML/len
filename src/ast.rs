use crate::lexer::{Token, TokenKind};
use crate::Int;
use chumsky::extra::ParserExtra;
use chumsky::pratt::{infix, left, postfix, prefix};
use chumsky::prelude::*;
use chumsky::Parser;
use std::borrow::Cow;
use std::collections::HashMap;
use std::ops::Not;
use tracing::span::Id;
use tracing::{debug, trace};
use tracing_test::traced_test;

#[derive(Debug, Clone)]
pub enum Ast {
    Expr(Expr),
    Binding { lhs: Identifier, rhs: Expr },
    Todo,
}

#[derive(Debug, Clone)]
pub enum Expr {
    Literal(Literal),
    FunctionCall(FunctionCall),
    Identifier(Identifier),
    Product(HashMap<String, Expr>),
    Lambda(Lambda),
    Todo,
}

#[derive(Debug, Clone)]
pub enum Literal {
    Integer(Int),
    String(String),
    Boolean(bool),
}

#[derive(Debug, Clone)]
pub struct FunctionCall {
    pub(crate) function: Box<Expr>,
    pub(crate) argument: Box<Expr>,
}

#[derive(Debug, Clone)]
pub struct Identifier {
    pub(crate) name: String,
}

#[derive(Debug, Clone)]
pub struct Lambda {
    pub from: Identifier,
    pub to: Box<Expr>,
}

#[derive(Debug, PartialEq)]
pub struct Spanned2<T>(T, SimpleSpan<usize>);

pub fn expression_parser<'s, E: ParserExtra<'s, &'s [TokenKind<'s>]>>(
) -> impl Parser<'s, &'s [TokenKind<'s>], Expr, extra::Err<Rich<'s, TokenKind<'s>>>> + Clone {
    recursive(|expr| {
        let literal = select! {
            TokenKind::Ident("true") => Literal::Boolean(true),
            TokenKind::Ident("false") => Literal::Boolean(false),
            TokenKind::Number(x) => Literal::Integer(x),
            TokenKind::String(x) => Literal::String(x.to_string()),
        }
        .map(Expr::Literal);

        let ident = select! {
            TokenKind::Ident(s) => Identifier { name: s.to_string() },
        }
        .map(Expr::Identifier);

        let grouping = expr.clone().delimited_by(
            just(TokenKind::LeftParenthesis),
            just(TokenKind::RightParenthesis),
        );

        let struct_elem = select! {
            TokenKind::Ident(s) => s.to_string(),
        }
        .then_ignore(just(TokenKind::Colon))
        .then(expr.clone());

        let r#struct = struct_elem
            .separated_by(just(TokenKind::Comma))
            .collect::<HashMap<_, _>>()
            .map(Expr::Product)
            .delimited_by(just(TokenKind::LeftCurly), just(TokenKind::RightCurly))
            .labelled("struct");

        let lambda = select! {
            TokenKind::Ident(s) => Identifier { name: s.to_string() },
        }
        .then_ignore(just(TokenKind::Arrow))
        .then(expr)
        .map(|(from, to)| Lambda {
            from,
            to: Box::new(to),
        })
        .map(Expr::Lambda)
        .labelled("lambda");

        let atom = choice((literal, r#struct, lambda, ident, grouping)).labelled("atom");

        // Left associative application
        let application = atom.clone().foldl(atom.repeated(), |op, o| {
            Expr::FunctionCall(FunctionCall {
                function: Box::new(op),
                argument: Box::new(o),
            })
        });

        let any_symbol = select! {
            TokenKind::Symbol(s) => Identifier { name: s.to_string() },
        }
        .map(Expr::Identifier);

        let mk_symbol = |c| {
            select! {
                TokenKind::Symbol(s) if c == s => Identifier { name: s.to_string() },
            }
            .map(Expr::Identifier)
        };

        let infix_fold = |left: Expr, op: Expr, right: Expr| {
            trace!("Creating infix");
            let first_op = Expr::FunctionCall(FunctionCall {
                function: Box::new(op),
                argument: Box::new(left),
            });
            Expr::FunctionCall(FunctionCall {
                function: Box::new(first_op),
                argument: Box::new(right),
            })
        };

        application.pratt((
            infix(left(2), any_symbol, infix_fold),
            infix(left(1), mk_symbol("+"), infix_fold),
            infix(left(1), mk_symbol("-"), infix_fold),
            infix(left(0), mk_symbol("$"), infix_fold),
        ))
    })
}

pub fn ast_parser<'s, E: ParserExtra<'s, &'s [TokenKind<'s>]>>(
) -> impl Parser<'s, &'s [TokenKind<'s>], Ast, extra::Err<Rich<'s, TokenKind<'s>>>> {
    let ep = expression_parser::<E>();

    let binding = select! {
        TokenKind::Ident(s) => Identifier { name: s.to_string() },
    }
    .then_ignore(just(TokenKind::Bind))
    .then(ep.clone())
    .map(|(lhs, rhs)| Ast::Binding { lhs, rhs });

    choice((binding, ep.map(Ast::Expr)))
}

#[cfg(test)]
mod tests {
    use super::*;
    use chumsky::error;
    use insta::assert_debug_snapshot;
    use rstest::rstest;
    use tracing_test::traced_test;
    type TestExtra = extra::Err<error::Cheap>;

    #[rstest]
    #[traced_test]
    fn test_expression<'src>(
        #[values(
            ("empty", &[][..]),
            ("ident", &[TokenKind::Ident("foo")][..]),
            ("int", &[TokenKind::Number(1.into())][..]),
            ("app", &[
                TokenKind::Ident("foo"),
                TokenKind::Number(1.into())
            ][..]),
            ("app_multiple", &[
                TokenKind::Ident("foo"),
                TokenKind::Ident("foo"),
                TokenKind::Ident("foo"),
                TokenKind::Ident("foo"),
            ][..]),
            ("grouping", &[
                TokenKind::LeftParenthesis,
                TokenKind::Ident("foo"),
                TokenKind::RightParenthesis,
            ][..]),
            ("bad-grouping", &[
                TokenKind::LeftParenthesis,
                TokenKind::Ident("foo"),
            ][..]),
            ("grouping_app", &[
                TokenKind::Ident("foo"),
                TokenKind::LeftParenthesis,
                TokenKind::Ident("bar"),
                TokenKind::Ident("baz"),
                TokenKind::RightParenthesis,
            ][..]),
            ("grouping_app_alt", &[
                TokenKind::LeftParenthesis,
                TokenKind::Ident("foo"),
                TokenKind::Ident("bar"),
                TokenKind::RightParenthesis,
                TokenKind::Ident("baz"),
            ][..]),
            ("infix", &[
                TokenKind::Number(1.into()),
                TokenKind::Symbol("+"),
                TokenKind::Number(1.into()),
            ][..]),
            ("infix_grouping", &[
                TokenKind::LeftParenthesis,
                TokenKind::Ident("a"),
                TokenKind::Symbol("+"),
                TokenKind::Ident("c"),
                TokenKind::RightParenthesis,
                TokenKind::Symbol("+"),
                TokenKind::LeftParenthesis,
                TokenKind::Ident("a"),
                TokenKind::Symbol("+"),
                TokenKind::Ident("c"),
                TokenKind::RightParenthesis,
            ][..]),
            ("infix_assoc", &[
                TokenKind::Ident("a"),
                TokenKind::Symbol("+"),
                TokenKind::Ident("b"),
                TokenKind::Symbol("*"),
                TokenKind::Ident("c"),
            ][..]),
            ("struct_simple", &[
                TokenKind::LeftCurly,
                TokenKind::Ident("a"),
                TokenKind::Colon,
                TokenKind::Ident("b"),
                TokenKind::RightCurly
            ][..]),
        )]
        input: (&str, &[TokenKind<'src>]),
    ) {
        let p = expression_parser::<TestExtra>();

        assert_debug_snapshot!(input.0, p.parse(input.1));
    }

    #[rstest]
    #[traced_test]
    fn test_ast<'src>(
        #[values(
            ("empty", &[][..]),
            ("assign", &[
                TokenKind::Ident("a"),
                TokenKind::Bind,
                TokenKind::Ident("b"),
            ][..]),
        )]
        input: (&str, &[TokenKind<'src>]),
    ) {
        let p = ast_parser::<TestExtra>();

        assert_debug_snapshot!(input.0, p.parse(input.1));
    }
}
