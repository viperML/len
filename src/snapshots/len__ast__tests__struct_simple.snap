---
source: src/ast.rs
expression: p.parse(input.1)
---
ParseResult {
    output: Some(
        Product(
            {
                "a": Identifier(
                    Identifier {
                        name: "b",
                    },
                ),
            },
        ),
    ),
    errs: [],
}
